From 0c40095c188d162be3fbc001566d26e7c38d20e1 Mon Sep 17 00:00:00 2001
From: Shinobu Uehara <shinobu.uehara.xc@renesas.com>
Date: Tue, 6 Aug 2013 17:20:49 +0900
Subject: [PATCH 1076/1083] mmc: tmio: Add disable auto CMD12

Disable auto CMD12 at IO_RW_EXTENDED multiple block transfer.

Signed-off-by: Shinobu Uehara <shinobu.uehara.xc@renesas.com>
---
 drivers/mmc/host/tmio_mmc_pio.c | 8 ++++++++
 include/linux/mfd/tmio.h        | 1 +
 2 files changed, 9 insertions(+)

diff --git a/drivers/mmc/host/tmio_mmc_pio.c b/drivers/mmc/host/tmio_mmc_pio.c
index 3f3b9a2..f2c413c 100644
--- a/drivers/mmc/host/tmio_mmc_pio.c
+++ b/drivers/mmc/host/tmio_mmc_pio.c
@@ -39,6 +39,7 @@
 #include <linux/mmc/mmc.h>
 #include <linux/mmc/slot-gpio.h>
 #include <linux/mmc/tmio.h>
+#include <linux/mmc/sdio.h>
 #include <linux/module.h>
 #include <linux/pagemap.h>
 #include <linux/platform_device.h>
@@ -317,6 +318,7 @@ static int tmio_mmc_start_command(struct tmio_mmc_host *host, struct mmc_command
 	struct mmc_data *data = host->data;
 	int c = cmd->opcode;
 	u32 irq_mask = TMIO_MASK_CMD;
+	struct tmio_mmc_data *pdata = host->pdata;
 
 	/* CMD12 is handled by hardware */
 	if (cmd->opcode == MMC_STOP_TRANSMISSION && !cmd->arg) {
@@ -347,6 +349,12 @@ static int tmio_mmc_start_command(struct tmio_mmc_host *host, struct mmc_command
 		if (data->blocks > 1) {
 			sd_ctrl_write16(host, CTL_STOP_INTERNAL_ACTION, 0x100);
 			c |= TRANSFER_MULTI;
+			if (cmd->opcode == SD_IO_RW_EXTENDED) {
+				/* Disable auto CMD12 at IO_RW_EXTENDED
+				  multiple block transfer */
+				if (pdata->disable_auto_cmd12)
+					pdata->disable_auto_cmd12(&c);
+			}
 		}
 		if (data->flags & MMC_DATA_READ)
 			c |= TRANSFER_READ;
diff --git a/include/linux/mfd/tmio.h b/include/linux/mfd/tmio.h
index d4c3303..89644ce 100644
--- a/include/linux/mfd/tmio.h
+++ b/include/linux/mfd/tmio.h
@@ -132,6 +132,7 @@ struct tmio_mmc_data {
 	int (*get_ro)(struct platform_device *host);
 	int (*write16_hook)(struct tmio_mmc_host *host, int addr);
 	bool (*dma_filter)(struct dma_chan *chan, void *arg);
+	void (*disable_auto_cmd12)(int *val);
 	/* clock management callbacks */
 	int (*clk_enable)(struct platform_device *pdev, unsigned int *f);
 	void (*clk_disable)(struct platform_device *pdev);
-- 
1.8.3.2

